### PR TITLE
Feat/validate, made validations function for model 

### PR DESCRIPTION
**Pull request by lsig**
validation function made and type annotated, need to make evaluate model function in metrics.py and also would be nice to extract get_prediction into a utils file since the function is used in both validate and train.

### CODE REVIEW

File: training/validate.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from transformers import BertTokenizer, PreTrainedModel
+
+device = torch.device("cpu:0" if torch.cuda.is_available() else "cpu")
+
+def get_prediction(model: PreTrainedModel, batch: DataLoader[tuple[torch.Tensor, torch.Tensor]]) -> PreTrainedModel:
```
Comment (njallskarp) [inline training/validate.py:8]: This is the same function, let's reuse. Possibly by having get_prediction, validate, and train all in the same file.

Also, is PreTrainedModel the right type declaration, I'm wondering if there is a similar type that is more indicative of what the function is actually outputting.. i.e. prediction output

**Review by njallskarp — CHANGES_REQUESTED** [review]
Added one comment  on get_prediction

Looks good but lets restructure code because of D.R.Y.

File: training/validate.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            outputs = get_prediction(model, batch)
+            loss = outputs[0]
+            # find the total loss
+            total_loss += loss.item()
+
+        pbar.set_postfix({'Batch': batch_idx+1, 'Loss': round(loss.item(),3)}, refresh=True)
+
+    total_loss /= len(val_loader)
+
+    #TODO: need to make evaluate_model function
```
Comment (njallskarp) [inline training/validate.py:37]: Remove this TODO

File: training/validate.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  input_ids = batch['input_ids'].to(device)
+  a_mask = batch['attention_mask'].to(device)
+  start_pos = batch['start_positions'].to(device)
+  end_pos = batch['end_positions'].to(device)
+        
+  return model(input_ids, attention_mask=a_mask, start_positions=start_pos, end_positions=end_pos)
+
+def validate(model: PreTrainedModel, tokenizer: BertTokenizer, val_loader: DataLoader, val_texts: list[str], val_questions: list[str], val_answers: list[dict]):
+
```
Comment (njallskarp) [inline training/validate.py:18]: Add """ """ multiline comment documenting function

File: training/validate.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from transformers import BertTokenizer, PreTrainedModel
+
+device = torch.device("cpu:0" if torch.cuda.is_available() else "cpu")
+
+def get_prediction(model: PreTrainedModel, batch: DataLoader[tuple[torch.Tensor, torch.Tensor]]) -> PreTrainedModel: 
+
```
Comment (njallskarp) [inline training/validate.py:9]: comment """ """ here too

File: training/validate.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from transformers import BertTokenizer, PreTrainedModel
+
+device = torch.device("cpu:0" if torch.cuda.is_available() else "cpu")
```
Comment (njallskarp) [inline training/validate.py:6]: Let's take this out of global scope and put it under `get_rediction`

**Review by njallskarp — CHANGES_REQUESTED** [review]
Adding request for changes

File: training/validate.py (hunk 1 above)
Comment (njallskarp) [reply training/validate.py:8]: Check out docs for the forward method here: https://huggingface.co/transformers/v3.0.2/model_doc/bert.html#transformers.BertForQuestionAnswering.forward

Return type is `tuple(torch.FloatTensor)`

File: training/validate.py (hunk 2 above)
Comment (lsig) [reply training/validate.py:37]: This function is something that still needs to be implemented, will remove when that work has finished. 💯

File: training/utils.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import torch
+from torch import device
+from torch.utils.data import DataLoader
+from transformers import BertForQuestionAnswering
+
+
+def get_rediction() -> device:
```
Comment (eysta00) [inline training/utils.py:7]: This should possible be a private function; i.e. __get_rediction() -> device:

**Review by eysta00 — APPROVED** [review]
looks good overall

File: types/declarations.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from typing import TypedDict
+
+
+class Answers(TypedDict): 
+    answers_end: int
```
Comment (njallskarp) [inline types/declarations.py:5]: Add a """ """ explaining the type

**Review by njallskarp — CHANGES_REQUESTED** [review]

**Review by njallskarp — APPROVED** [review]

**Pull request merged**